components/camel-grpc/src/test/java/org/apache/camel/component/grpc/RouteControlledStreamObserverTest.java

```diff
@@ -0,0 +1,242 @@
+        requestObserver.onNext(pingRequest);
+        requestObserver.onNext(pingRequest);
+        requestObserver.onCompleted();
+        latch.await(5, TimeUnit.SECONDS);
+
+        PongResponse pongResponse = responseObserver.getPongResponse();
+
+        assertNotNull(pongResponse);
+        assertEquals(GRPC_TEST_PING_ID, pongResponse.getPongId());
+        assertEquals(GRPC_TEST_PING_VALUE + GRPC_TEST_PONG_VALUE, pongResponse.getPongName());
+    }
+
+    @Test
+    public void testAsyncAsyncMethodInAsync() throws Exception {
+        LOG.info("gRPC pingAsyncAsync method async test start");
+        final CountDownLatch latch = new CountDownLatch(1);
+        PingRequest pingRequest
+                = PingRequest.newBuilder().setPingName(GRPC_TEST_PING_VALUE).setPingId(GRPC_TEST_PING_ID).build();
+        PongResponseStreamObserver responseObserver = new PongResponseStreamObserver(latch);
+
+        StreamObserver<PingRequest> requestObserver = asyncNonBlockingStub.pingAsyncAsync(responseObserver);
+        requestObserver.onNext(pingRequest);
+        requestObserver.onNext(pingRequest);
+        requestObserver.onCompleted();
+        latch.await(5, TimeUnit.SECONDS);
+
+        PongResponse pongResponse = responseObserver.getPongResponse();
+
+        assertNotNull(pongResponse);
+        assertEquals(GRPC_TEST_PING_ID, pongResponse.getPongId());
+        assertEquals(GRPC_TEST_PING_VALUE + GRPC_TEST_PONG_VALUE, pongResponse.getPongName());
+    }
+
+    @Override
+    protected RouteBuilder createRouteBuilder() {
+        return new RouteBuilder() {
+            private void process(Exchange exchange) {
+                Message message = exchange.getIn();
+                PingRequest pingRequest = message.getBody(PingRequest.class);
+                StreamObserver<Object> responseObserver = (StreamObserver<Object>) exchange.getProperty("responseObserver");
+                PongResponse pongResponse
+                        = PongResponse.newBuilder().setPongName(pingRequest.getPingName() + GRPC_TEST_PONG_VALUE)
+                                .setPongId(pingRequest.getPingId()).build();
+                message.setBody(pongResponse, PongResponse.class);
+                exchange.setMessage(message);
+                responseObserver.onNext(pongResponse);
+                responseObserver.onCompleted();
+            }
+
+            @Override
+            public void configure() {
+                from("grpc://localhost:" + GRPC_SYNC_REQUEST_TEST_PORT
+                     + "/org.apache.camel.component.grpc.PingPong?synchronous=true&consumerStrategy=PROPAGATION&routeControlledStreamObserver=true")
+                             .process(this::process);
+
+                from("grpc://localhost:" + GRPC_ASYNC_REQUEST_TEST_PORT
+                     + "/org.apache.camel.component.grpc.PingPong?synchronous=true&consumerStrategy=AGGREGATION")
```

Comment:
Seems to routeControlledStreamObserver=true and consumerStrategy=AGGREGATION not compatible. We should notify that option will be ignored and not applied or just fail route start as it cause unpredictable result.

Reply:
Agree. Included a failure test case and added a check in the GrpcConsumer to prevent server start event in case of incorrect configuration.